elements/howto-label/howto-label.js

```diff
@@ -0,0 +1,142 @@
+    }
+
+    constructor() {
+      super();
+      this.attachShadow({mode: 'open'});
+      this.shadowRoot.appendChild(template.content.cloneNode(true));
+      // Listen for the slotchange event to label any wrapped children
+      this.shadowRoot.querySelector('slot')
+        .addEventListener('slotchange', this._forChanged.bind(this));
+      this.addEventListener('click', this._onClick.bind(this));
+    }
+
+    /**
+     * connectedCallback() will run once the element is attached to the document.
+     * It's the earliest point at which you may touch attributes, and is a good
+     * time to auto-generate an id for the element if it does not already have
+     * one. The labelled element will use this id to refer to the label with
+     * aria-labelledby.
+     */
+    connectedCallback() {
+      if (!this.id) {
+        this.id = `howto-label-generated-${howtoLabelCounter++}`;
+      }
+
+      // Fire _forChanged at startup to label any wrapped elements.
+      this._forChanged();
+    }
+
+    get for() {
+      const value = this.getAttribute('for');
+      return value === null ? '' : value;
+    }
+
+    set for(value) {
+      this.setAttribute('for', value);
+    }
+
+    attributeChangedCallback(name, oldVal, newVal) {
+      if (name === 'for') {
+        this._forChanged();
+      }
+    }
+
+    /**
+     * Remove any previous labels.
+     * Find the new element to label.
+     * Label it.
+     */
+    _forChanged() {
+      let target = this._currentLabelTarget();
+      if (target) {
+        target.removeAttribute('aria-labelledby');
+      }
+      target = this._findTarget();
+      target.setAttribute('aria-labelledby', this.id);
+    }
+
+    _onClick(event) {
+      let el = this._currentLabelTarget();
+      // If nothing is labeled, or if a wrapped child was clicked on, return.
+      if (!el || event.target === el) {
+        return;
+      }
+      el.focus();
+      el.click();
+    }
+
+    _currentLabelTarget() {
+      // Query inside of the current scope. This could either be a shadow root
+      // or the main document.
+      let scope = this.getRootNode();
+      return scope.querySelector(`[aria-labelledby="${this.id}"]`);
+    }
+
+    /**
+     * If there is a for property, return the element with that id.
+     * Else, return the explicitly labeled child.
+     * Else, return the first element child (assume it's implicitly labeled).
+     */
+    _findTarget() {
+      if (this.for) {
+        // external target
+        let scope = this.getRootNode();
+        return scope.getElementById(this.for);
+      } else {
+        let slottedChildren = this.shadowRoot
+          .querySelector('slot').assignedNodes({flatten: true});
+        let el;
+        slottedChildren.forEach(child => {
+          // Ignore text nodes
+          if (child.nodeType === Node.TEXT_NODE) {
+            return;
+          }
+          // Implicit internal target, first element child
+          if (!el) {
```

Comment:
This `if` is a bit confusing. Couldn’t you do it like this?

```js
// Get all non-text slotted children
let slottedChildren = 
  this._slot.assignedNodes({flatten: true})
    .filter(child => child.nodeType !== Node.TEXT_NODE);
// Find the first one that defines an explicit external target
let el =
  slottedChildren
    .find(child => child.hasAttribute('howto-label-target'));
// Return that first explicit external target, or the first child if there is none.
return el || slottedChildren[0];
``

Reply:
yeah but now you're looping through the children twice :P